core/src/main/java/io/grpc/ClientInterceptors.java

```diff
@@ -204,4 +216,190 @@ public final void start(Listener<RespT> responseListener, Metadata headers) {
       }
     }
   }
+
+  public static class RetryingInterceptor implements ClientInterceptor {
+    // TODO(lukaszx0) have default logger for this one to give more visibility into when/what gets
+    // retried?
+    // private final Logger logger = Logger.getLogger(RetryingInterceptor.class.getName());
+    private final RetryPolicy retryPolicy;
+
+    public RetryingInterceptor(RetryPolicy retryPolicy) {
+      this.retryPolicy = retryPolicy;
+    }
+
+    public RetryingInterceptor() {
+      this.retryPolicy = new IdempotentRetryPolicy();
+    }
+
+    @Override
+    public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+            MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+      return new RetryingCall<ReqT, RespT>(retryPolicy, method, callOptions, next);
+    }
+
+    private class RetryingCall<ReqT, RespT> extends ClientCall<ReqT, RespT> {
+      private final RetryPolicy retryPolicy;
+      private final MethodDescriptor<ReqT, RespT> method;
+      private final CallOptions callOptions;
+      private final Channel channel;
+      private final ScheduledExecutorService scheduledExecutor;
+      private Listener<RespT> responseListener;
+      private Metadata requestHeaders;
+      private ReqT requestMessage;
+      private boolean compressionEnabled;
+      private final Queue<AttemptListener> attemptListeners =
+              new ConcurrentLinkedQueue<AttemptListener>();
+      private volatile AttemptListener latestResponse;
+      private volatile ScheduledFuture<?> retryTask;
+
+      RetryingCall(RetryPolicy retryPolicy, MethodDescriptor<ReqT, RespT> method,
+                   CallOptions callOptions, Channel channel) {
+        this.method = method;
+        this.callOptions = callOptions;
+        this.channel = channel;
+        this.retryPolicy = retryPolicy;
+        this.scheduledExecutor = SharedResourceHolder.get(TIMER_SERVICE);
```

Comment:
@ejona86 Since this is refcounted, I think I need to `release` it once I'm done using it, right? Maybe `useResponse` would be the best place to do it?